clients/src/main/java/org/apache/kafka/clients/ApiVersions.java

```diff
@@ -34,10 +34,21 @@ public class ApiVersions {
 
     private final Map<String, NodeApiVersions> nodeApiVersions = new HashMap<>();
     private byte maxUsableProduceMagic = RecordBatch.CURRENT_MAGIC_VALUE;
+    private short maxProduceSupportedVersion = ApiKeys.PRODUCE.latestVersion();
 
     public synchronized void update(String nodeId, NodeApiVersions nodeApiVersions) {
         this.nodeApiVersions.put(nodeId, nodeApiVersions);
         this.maxUsableProduceMagic = computeMaxUsableProduceMagic();
+        this.maxProduceSupportedVersion = computeMaxProduceSupportedVersion();
+    }
+
+    private short computeMaxProduceSupportedVersion() {
+        Optional<Short> knownBrokerNodesMinSupportedVersionForProduce = this.nodeApiVersions.values().stream()
```

Comment:
I wonder if we should also confirm that the IBP is at least 2.8 on the cluster. (Note this only applies to ZK clusters as all kraft clusters are 3.0 and above/have topic IDs)

There is a case someone is running a new cluster with a very old IBP.

This check would be similar to what `!topicIds.entrySet().stream().anyMatch(e -> e.getValue() == Uuid.ZERO_UUID)` was doing before.

Reply:
> I wonder if we should also confirm that the IBP is at least 2.8 on the cluster. (Note this only applies to ZK clusters as all kraft clusters are 3.0 and above/have topic IDs)

I don't think the client access the IBP of the broker. It only access the min/max supported versions of a request which IBP is just a proxy for. Am I missing a route on client side that access IBP versions?

I use `!topicIds.entrySet().stream().anyMatch(e -> e.getValue() == Uuid.ZERO_UUID)` to make sure the `ProducerMetadata` as well is a topic id aware as this is another request that we relay on to map between topic and topic ids

Reply:
My understanding is that ApiVersions also contains features including metadata version :)
I believe if the cluster supports this api, it will also support the api versions that returns this. But might be a bit complicated 

I wrote this before I saw the uuid check so maybe this is ok. I just wonder if it will be slow to check all the topic IDs each time.

Reply:
We only check ids for topic included in the batch, and we catch these from the metadata which at this point the producer already fetched them or cached them.

Reply:
sounds good. this is consistent with the fetching logic 👍